# Final exam (Individual)


### <p style="text-align: right;"> &#9989; Siddak Marwaha.</p>
### <p style="text-align: right;"> &#9989; siddakmarwaha.</p>

# CMSE 202 Final (Fall 2022)

The goal of this final is to give you the opportunity to test out some of the skills that you've developed having now finished CMSE 202. In particular, you'll be committing and pushing repository changes to a GitHub repository, working with data to build a network graph, performing regression analysis, and classifying data using a machine learning classifier. You should find that you have all of the skills necessary to complete this exam having completed the second half of CMSE 202!

You are encouraged to look through the entire exam before you get started so that you can appropriately budget your time and understand the broad goals of the exam. Once you've read through it, you'll probably want to make sure you do Part 1 first to ensure that your GitHub repository is working correctly. Let your instructor know right away if you run into issues!

The exam is set up so that even if you get stuck on one part there are opportunities to get points on the other parts, so consider jumping ahead if you feel like you aren't making progress and then come back later if you have time.

If you have any questions during the exam, you may ask the instructor, the TA, or the LA privately. If you are attending in-person, simply raise your hand and one of us will come over to you. If you are attending via-Zoom, please use the ask-for-help feature or privately message the instructor, the TA, or the LA. We will make any announcements or exam clarifications/corrections to the class via the following Google Document.

[LINK](https://docs.google.com/document/d/16x_nbg8coBwf9hyLXRxOXAvrugkRhjeaBQJMiUwdBw8/edit?usp=sharing)

**Important note about using online resources**: This exam is "open internet". That means that you can look up documentation, google how to accomplish certain Python tasks, etc. Being able to effectively use the internet for computational modeling and data science is a very important skill, so we want to make sure you have the opportunity to exercise that skill. **However**: The use of any person-to-person communication software is absolutely not acceptable. If you are seen accessing your email, using a chat program (e.g. Slack), or any sort of collaborative cloud storage or document software (e.g. Google Documents other than the one we will make announcements on), you will be at risk for receiving a zero on the exam.

**Keep your eyes on your screen!** This exam is designed to give *you* the opportunity to show the instructor what you can do and you should hold yourself accountable for maintaining a high level of academic integrity. If any of the instructors observe suspicious behavior, you will, again, risk receiving a zero.

------
<a id="toc"></a>

## Table of contents


0. [Part 0: Upgrade packages](#part0) (1 point)

1. [Part 1: Git](#part1) (9 points)

2. [Part 2: Data Preprocessing](#part2) (20 points)

3. [Part 3: Regression](#part3) (15 points)

4. [Part 4: Principal Component Analysis](#part4) (12 points)

5. [Part 5: Support Vector Machines](#part5) (15 points)

6. [Part 6: Regression Redux](#part6) (7 points)

7. [Part 7: Conceptual Questions](#part7) (14 points)

8. [Part 8: Conclusion](#conclusion) (3 points)

In [1]:
grades = [1, 9, 20, 15, 12, 15, 7, 14, 3]

print(f"The total grade for this final is {sum(grades)}" )

The total grade for this final is 96


---
<a id="part0"></a>
[Back to ToC](#toc)

# Part 0: Upgrade Packages

**&#9989; Question 0.1 (1 point)**: Run the cell below. Do you have the correct packages ? If not upgrade them. **You must do this in order to avoid issues in the rest of the notebook.**


In [2]:

# Standard libraries
%pylab
%matplotlib inline

import seaborn as sns
import pandas as pd

# Data cleaning Libraries
from sklearn.decomposition import PCA
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.preprocessing import PowerTransformer
from sklearn.utils import resample
# Classifier Libraries
from sklearn.linear_model import Perceptron
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix, accuracy_score, ConfusionMatrixDisplay
from sklearn.svm import SVC, LinearSVC

import statsmodels.api as sm 

from sklearn import __version__ as sk_version
sns.set_context("talk")

print(f"Sklearn version should be 1.1.3 and I have {sk_version}")


Using matplotlib backend: agg
Populating the interactive namespace from numpy and matplotlib
Sklearn version should be 1.1.3 and I have 1.1.3


---
<a id="part1"></a>
[Back to ToC](#toc)

## Part 1: Git (9 points)

For this assignment, you're going to add it to the `cmse202-f22-turnin` repository you created in class so that you can track your progress on the assignment and preserve the final version that you turn in. In order to do this you need to

**&#9989; Question 1.1 (1 point)**: Navigate to your `cmse202-f22-turnin` **local** repository and create a new directory called `final` and copy this notebook in that new directory.

``` bash
mkdir final

```


&#9989; **Question 1.2 (3 points)** Check the status of your local `git`.

``` bash 
git status 

```
Copy and paste below the output of the command.

``` bash
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
        modified:   ../hw-01 (modified content, untracked content)
        modified:   ../hw-03/HW-03_ABMAndGraphTheory-STUDENT.ipynb

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        ../.ipynb_checkpoints/
        ../CMSE202_Final-STUDENT.ipynb
        ../hw-02/.ipynb_checkpoints/HW-02_PandasAndObjects-STUDENT (2)-checkpoint.ipynb
        ../hw-02/HW-02_PandasAndObjects-STUDENT (2).ipynb
        ../hw-02/netflix_titles.csv
        ../hw-03/.ipynb_checkpoints/
        ../hw-03/asoiaf-book1-edges.csv
        ../hw-03/store_placement.py
        ../hw-04/.ipynb_checkpoints/
        ../hw-04/ads.csv
        ../hw-04/housing.csv
        ../midterm/.ipynb_checkpoints/
        ../midterm/energy_dataset.xlsx
        ../midterm/psi.dat

no changes added to commit (use "git add" and/or "git commit -a")
```

What is the name of the branch you are in ? 

``` bash
master
```

**Important:** You should be in the `main` branch. If you are not switch to the `main` branch

&#9989; **Question 1.3 (3 points):**
Add your name and GitHub username to the top of the notebook, then add and commit **ONLY** the notebook.

``` bash
git add CMSE202_Final-STUDENT.ipynb
git commit -m "Initial Commit" 
```

What is the commit message you used ?

``` bash
Initial Commit
```

&#9989; **Question 1.4 (1 point):** Before moving on. Check that the notebook you are working on is the correct one. Run the following cell. **Are you in the new folder you just created?** If not close this notebook and open the one in the `final` folder


In [3]:
!pwd

/home/marwaha3/cmse202-f22-turnin/final


&#9989; **Question 1.5 (1 point):** Finally push the updated notebook to GitHub.

``` bash
git push https://github.com/siddakmarwaha/cmse202-f22-turnin.git
```


**Important**: Double check you've added your Professor and your TA as collaborators to your "turnin" repository (you should have done this in the previous homework assignment).

**Also important**: Make sure that the version of this notebook that you are working on is the same one that you just added to your repository! If you are working on a different copy of the notebook, **none of your changes will be tracked**!

If everything went as intended, the file should now show up on your GitHub account in the "`cmse202-f22-turnin`" repository inside the `final` directory that you just created.


---
<a id="part2"></a>
[Back to ToC](#toc)

# Part 2. Data preprocessing (19 points)

For this assignment we’re going to be working with a dataset that contains measurements of the characteristics of multiple wines. This includes things like the acidity and density. You’ll be asked to use this information, along with the machine learning tools that we’ve used in class, to determine whether the wine type is Red or White.

The dataset is located at:
`https://raw.githubusercontent.com/msu-cmse-courses/cmse202-F22-data/main/data/winequality.csv`


**&#9989; Question 2.1 (4 points):** Do this

1. Download the data and type the command you used to download the data

2. Read the `winequality.csv` file into a dataframe

3. Print out the **unique** labels in the `type` column. 

4. Display the first 10 rows of the dataset.

**Note**: each row represents one data point and each column (except the `type` column) represents one feature. The `type` column corresponds to the class labels for every data point. There are two types of unique class labels in the `type` column. 

In [4]:
# Put the command you used to download the data here
winequality = pd.read_csv('winequality.csv')
print(winequality['type'].unique())


['white' 'red']


In [5]:
# Put your code here
winequality.head(10)

,type,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,white,7.0,0.27,0.36,20.7,0.045,45.0,170.0,1.0010,3.00,0.45,8.8,6
1,white,6.3,0.30,0.34,1.6,0.049,14.0,132.0,0.9940,3.30,0.49,9.5,6
2,white,8.1,0.28,0.40,6.9,0.050,30.0,97.0,0.9951,3.26,0.44,10.1,6
3,white,7.2,0.23,0.32,8.5,0.058,47.0,186.0,0.9956,3.19,0.40,9.9,6
4,white,7.2,0.23,0.32,8.5,0.058,47.0,186.0,0.9956,3.19,0.40,9.9,6
5,white,8.1,0.28,0.40,6.9,0.050,30.0,97.0,0.9951,3.26,0.44,10.1,6
6,white,6.2,0.32,0.16,7.0,0.045,30.0,136.0,0.9949,3.18,0.47,9.6,6
7,white,7.0,0.27,0.36,20.7,0.045,45.0,170.0,1.0010,3.00,0.45,8.8,6
8,white,6.3,0.30,0.34,1.6,0.049,14.0,132.0,0.9940,3.30,0.49,9.5,6
9,white,8.1,0.22,0.43,1.5,0.044,28.0,129.0,0.9938,3.22,0.45,11.0,6


**&#9989; Question 2.2 (7 points):** Do the following:

1. Drop the `NaN` in the dataset

2. Drop the `quality` column

2. Replace all of the strings in your `type` column with integers based on the following:

    | original type | integer type |
    | -------- | -------- |
    | white | 0 |
    | red | 1 |

3. Once you've replaced the labels, display your DataFrame and confirm that it looks correct.

In [6]:
# Put your code here
winequality = winequality.dropna()
winequality = winequality.drop(columns='quality')

for i in range(len(winequality)):
    if winequality['type'].iloc[i]=='white':
        winequality['type'].iloc[i]=0
    else:
        winequality['type'].iloc[i]=1
winequality

<ipython-input-6-775726fbcc41>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  winequality['type'].iloc[i]=0
<ipython-input-6-775726fbcc41>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  winequality['type'].iloc[i]=1


,type,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol
0,0,7.0,0.270,0.36,20.7,0.045,45.0,170.0,1.00100,3.00,0.45,8.8
1,0,6.3,0.300,0.34,1.6,0.049,14.0,132.0,0.99400,3.30,0.49,9.5
2,0,8.1,0.280,0.40,6.9,0.050,30.0,97.0,0.99510,3.26,0.44,10.1
3,0,7.2,0.230,0.32,8.5,0.058,47.0,186.0,0.99560,3.19,0.40,9.9
4,0,7.2,0.230,0.32,8.5,0.058,47.0,186.0,0.99560,3.19,0.40,9.9
...,...,...,...,...,...,...,...,...,...,...,...,...
6491,1,6.8,0.620,0.08,1.9,0.068,28.0,38.0,0.99651,3.42,0.82,9.5
6492,1,6.2,0.600,0.08,2.0,0.090,32.0,44.0,0.99490,3.45,0.58,10.5
6494,1,6.3,0.510,0.13,2.3,0.076,29.0,40.0,0.99574,3.42,0.75,11.0
6495,1,5.9,0.645,0.12,2.0,0.075,32.0,44.0,0.99547,3.57,0.71,10.2


**&#9989; Question 2.3 (2 points):** As we've seen, when working with `scikit-learn` it can be much easier to work with the data if we have separate variables: one that stores the feature matrix and one that stores the class labels.

**Do This:** Split your DataFrame so that you have two separate DataFrames: (1) one called `features`, which contains all columns of features; and (2) one called `labels`, which is a single-column dataframe that contains all of the *new* integer labels you just created. 

In [7]:
# Put your code here
features = winequality.iloc[:,1:]
labels = pd.DataFrame(winequality.iloc[:,0])


**&#9989; Question 2.4 (1 points):** Do this:

We need to scale the data. This isn't something we've talked very much about in-class, so we don't expect you to know exactly what is happening here. But it will be important for our PCA later on in the assignment. We decided to put it here so you were aware of it's existence. 

**Run the following bit of code below.**

In [8]:
features = pd.DataFrame(PowerTransformer().fit_transform(features), 
                            columns=features.columns, 
                            index=features.index)

**&#9989; Question 2.5 (3 points):** Do this:

1. Split your data into a training and a testing set with a training set representing 75% of your data. For reproducibility , set the `random_state` argument to `314159`. 

2. Print the lengths to show you have the right number of entries for the training and testing sets.

In [9]:
X = features
y = labels
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25,random_state=314159)
len(X_train), len(X_test), len(y_train), len(y_test)

(4847, 1616, 4847, 1616)

### &#128721; STOP
**Pause to commit your changes to your Git repository! (2 points)**

Take a moment to save your notebook, commit the changes to your Git repository using the commit message "Committing Part 2", and push the changes to GitHub.

---

<a id="part3"></a>
[Back to ToC](#toc)

# Part 3. Regression (15 Points)

**&#9989; Question 3.1.1 (3 points):**
Let's start making some prediction. 

1. Make a Logistic Regression model using `statsmodels` for predicting the type of wine. 

2. Print out the summary of your model fits.


In [10]:
y_test = y_test.astype(int)
y_train = y_train.astype(int)


In [11]:
# Write your code here
logit_model = sm.Logit(y_train,X_train)
result = logit_model.fit()
print(result.summary())

         Current function value: 0.193646
         Iterations: 35
                           Logit Regression Results                           
Dep. Variable:                   type   No. Observations:                 4847
Model:                          Logit   Df Residuals:                     4837
Method:                           MLE   Df Model:                            9
Date:                Tue, 13 Dec 2022   Pseudo R-squ.:                  0.6530
Time:                        21:17:41   Log-Likelihood:                -938.60
converged:                      False   LL-Null:                       -2704.8
Covariance Type:            nonrobust   LLR p-value:                     0.000
                           coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------------
fixed acidity            1.4959      0.115     12.993      0.000       1.270       1.722
volatile acidity         1.3956    

/home/marwaha3/.local/lib/python3.8/site-packages/statsmodels/base/model.py:604: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


**&#9989; Question 3.1.2 (2 points):** Further examine the results by  

1. printing a classification report
2. making a confusion matrix

In [30]:
# Write your code here
predict_vectors = X_train
true_labels = y_train

print("Predicting names on the test set")
y_pred = result.predict()
for i in range(len(y_pred)):
    if y_pred[i]>=0.5:
        y_pred[i]=1
    else:
        y_pred[i]=0

print(classification_report(true_labels, y_pred))
print(confusion_matrix(true_labels, y_pred))


Predicting names on the test set
              precision    recall  f1-score   support

           0       1.00      0.90      0.95      3654
           1       0.77      0.99      0.86      1193

    accuracy                           0.92      4847
   macro avg       0.88      0.95      0.91      4847
weighted avg       0.94      0.92      0.93      4847

[[3294  360]
 [  11 1182]]


**&#9989; Question 3.2.1 (3 points):** Make an reduced Logistic Regression model for predicting the type of wine that uses just three parameters (two from the dataset + a constant). Print out the summary of your model fits.



In [48]:
#Write your code here
reduced_features = features[['volatile acidity', 'citric acid']]
X = reduced_features
y = labels
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25,random_state=314159)
len(X_train), len(X_test), len(y_train), len(y_test)

(4847, 1616, 4847, 1616)

In [49]:
y_test = y_test.astype(int)
y_train = y_train.astype(int)


In [50]:
logit_model = sm.Logit(y_train,sm.add_constant(X_train))
result = logit_model.fit()
print(result.summary())

Optimization terminated successfully.
         Current function value: 0.321510
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:                   type   No. Observations:                 4847
Model:                          Logit   Df Residuals:                     4844
Method:                           MLE   Df Model:                            2
Date:                Tue, 13 Dec 2022   Pseudo R-squ.:                  0.4239
Time:                        22:05:27   Log-Likelihood:                -1558.4
converged:                       True   LL-Null:                       -2704.8
Covariance Type:            nonrobust   LLR p-value:                     0.000
                       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------
const               -2.0963      0.065    -32.446      0.000      -2.223      -1.970
volatile ac

**&#9989; Question 3.2.2 (2 points):** Same as above, examine the results of your reduced model by printing a classification report and a confusion matrix

In [51]:
#Write your code here
predict_vectors = X_train
true_labels = y_train

print("Predicting names on the test set")
y_pred = result.predict()
for i in range(len(y_pred)):
    if y_pred[i]>=0.5:
        y_pred[i]=1
    else:
        y_pred[i]=0

print(classification_report(true_labels, y_pred))
print(confusion_matrix(true_labels, y_pred))


Predicting names on the test set
              precision    recall  f1-score   support

           0       0.90      0.93      0.91      3654
           1       0.77      0.67      0.71      1193

    accuracy                           0.87      4847
   macro avg       0.83      0.80      0.81      4847
weighted avg       0.86      0.87      0.87      4847

[[3415  239]
 [ 397  796]]


**&#9989; Question 3.3 (3 points):** How did you pick the best parameters?

*I chose the features that had the lowest p value, in this case 0.*

### &#128721; STOP
**Pause to commit your changes to your Git repository! (2 points)**

Take a moment to save your notebook, commit the changes to your Git repository using the commit message "Committing Part 3", and push the changes to GitHub.


---
<a id="part4"></a>
[Back to ToC](#toc)

# Part 4. Principal Component Analysis (12 points)

The full model uses all 12 features to predict the results. In many cases, we might need to see how close we can get with fewer features. Instead of simply removing features, we will use a Principal Component Analysis (PCA) to determine the combined features that contribute the most to the model (through their accounted variance).


**&#9989; Question 4.1 (5 points):** Run a Principle Component Analysis (PCA)

Since we only have 12 features to start with, let's see how well we can do if we try to aggressively reduce the feature count and use only **2** principle components. 

1. Using `PCA()` and the associated methods, run a principle component analysis on your `features` dataset using only 2 components. 

2. Transform the dataset into a new dataset and call it `pca_features`. 

3. Print the `explained_variance_ratio_` and its sum.

In [72]:
X = features
y = labels


In [73]:
# Put your code here
pca = PCA(n_components=2)
_ = pca.fit(features)

In [75]:
pca_features = pca.transform(features)

print("Training set changed from a size of: ", features.shape, ' to: ', pca_features.shape)


Training set changed from a size of:  (6463, 11)  to:  (6463, 2)


In [76]:
print(pca.explained_variance_ratio_, 'and sum:', sum(pca.explained_variance_ratio_))

[0.30496612 0.18394275] and sum: 0.4889088724339897


&#9989; **Question 4.1.1 (2 points):** What is the **total** explained variance ratio captured by the 2 principle components? (just quote the number) How well do you think a model with these many features will perform? Why?

<font size=+3>&#9998;</font> 0.4889088724339897. I think the model would not perform as well since the explained variance ratio's sum is even less than 0.5 so not enough.

**&#9989; Question 4.2 (3 points):** Do the following:

1. Split your new features (`pca_features`) and corresponding labels (the labels are the same as before) into a training and a testing set, with the training set representing 75% of your data. For reproducibility, set the `random_state` argument to `314159`. 

2. Print the lengths to show you have the right number of entries.

Basically do the same you have done in Q2.5 but with the `pca_features` dataset now.

In [77]:
# Put your code here
X = pca_features
y = labels
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25,random_state=314159)
len(X_train), len(X_test), len(y_train), len(y_test)

(4847, 1616, 4847, 1616)

---
### &#128721; STOP
**Pause to commit your changes to your Git repository! (2 points)**

Take a moment to save your notebook, commit the changes to your Git repository using the commit message "Committing Part 4".

---

---
<a id="part5"></a>
[Back to ToC](#toc)

# Part 5. Support Vector Machine (15 points)

Let's see how an SVM performs

**&#9989; Question 5.1 (9 points):** Do the following:

1. Build a `SVC` model with a `linear` kernel. (2 points), 

2. Use `GridSearchCV` to find the best `C` parameter from this list: `'C':[0.0001, 0.001, 0.1, 1, 10]`, (2 points)
  
3. **Fit** the above model on the training set and **print** the best estimator (2 points)

4. Use your best estimator on the test dataset to make prediction. (1 point)

5. Print a [classification report](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html), based on its performance on the testing dataset. (1 point)

6. Print the [confusion matrix](https://scikit-learn.org/stable/auto_examples/model_selection/plot_confusion_matrix.html#sphx-glr-auto-examples-model-selection-plot-confusion-matrix-py) on the testing dataset  (1 point)

**Note:** You can set the `GridSearchCV` keyword argument `n_jobs = -1` to speed things up (this allows the process to run on multiple cores.)

**Note:** If `GridSearchCV` is slow you can use the built-in model `LinearSVC` in `sklearn`. This is the same as `SVC(kernel = 'linear')` but it is faster. 

In [ ]:
# Put your code here


**&#9989; Question 5.1.1 (4 points):** Answer the following questions:

1. What is the accuracy score of your model ? **Do not use code** You should be able to read it from the output above!

2. How does your SVM model compare with the Regression models? 


---
### &#128721; STOP
**Pause to commit your changes to your Git repository! (2 points)**

Take a moment to save your notebook, commit the changes to your Git repository using the commit message "Committing Part 5".

---

<a id="part3"></a>
[Back to ToC](#toc)

# Part 6. Regression Redux (7 Points)

**&#9989; Question 6.1.1 (3 points):** Let's investigate the Logistic model again and see how it performs on the new PCA-transformed dataset. 

**Do this**: Make an Logistic model using the PCA features (2 features + constant). You should add a constant to the two PCA features, giving you three total features.


In [ ]:
#Write your code here

**&#9989; Question 6.1.2 (2 points):**  Examine the results by making a confusion matrix (2 Points)

In [ ]:
#Write your code here

---
### &#128721; STOP
**Pause to commit your changes to your Git repository! (2 points)**

Take a moment to save your notebook, commit the changes to your Git repository using the commit message "Committing Part 6".

---

---
<a id="part7"></a>
[Back to ToC](#toc)

# Part 7. Conceptual Questions (14 Points)

**&#9989; Question 7.1 (4 points):** Compare your results from 3.2.2 and 6.1.2, two Logistic regression models that used the three most significant features. Which one performed better? Why? Your explanation should discuss the difference between the features in the two models and how that difference might affect how well the model fits the data.  

*Write your response here*

#### 7.2 
The following questions probe your understanding of  `recall` and `precision`. You’ll be given a specific scenario, and you will need to decide whether to select a Machine Learning model that maximizes `recall` or `precision`.


**&#9989; Question 7.2 (4 points):**  A new disease has been detected in Sweden. Doctors have taken multiple measurements of patients and passed them along to you. You’ve tested various models to predict which patients have the new disease. The Swedish health experts have told you it is critical that they identify **all** of the patients with the new disease so they can put them into quarantine. They tell you that it doesn’t matter if your model accidentally flags some people as having the disease even when they don’t, as it’s much better to tell people who aren’t sick to quarantine than to tell a sick person they don’t need to quarantine.

Given this situation, should you choose the model that maximizes `recall` or the model that maximizes `precision`? Why?

*Write your response here*

**&#9989; Question 7.2.2 (4 points):**
You are working for the state of Michigan, helping them with their new unemployment insurance program. You’ve created various models that are meant to detect fraud–i.e., determine whether or not someone applied for unemployment insurance when they shouldn’t have. Your boss tells you that it’s critical that your model doesn’t accidentally accuse an innocent person of fraud. They say it’s fine if your model misses some people who committed fraud as long as there are no false accusations. 

Given this situation, should you choose the model that maximizes `recall` or the model that maximizes `precision`? Why?

*Write your response here*

---
### &#128721; STOP
**Pause to commit your changes to your Git repository!** (2 points)

Take a moment to save your notebook, commit the changes to your Git repository using the commit message "Committing Part 7".

---

---
<a id="conclusion"></a>
[Back to ToC](#toc)

# Part 8. Conclusion (3 points)

Make sure all of your changes to your repository are committed and pushed to GitHub. 
Before you leave

1. Have you added your name and github username at the top of this notebook? ? (1 point)

2. Push the changes to your GitHub repository (1 point)

3. Upload your notebook to D2L in case something went wrong with your repository or if you couldn't get the repository to work.  (1 point)

## You're done! Congrats on finishing your CMSE 202 Midterm!

In [ ]:

# FINAL GRADE (Instructor's use only)",
parts = [part_0, part_1, part_2, part_3, part_4, part_5, part_6, part_7]
print(f"Part 0: {part_0}/{grades[0]}")
print(f"Part 1: {part_1}/{grades[1]}")
print(f"Part 2: {part_2}/{grades[2]}")
print(f"Part 3: {part_3}/{grades[3]}")
print(f"Part 4: {part_4}/{grades[4]}")
print(f"Part 5: {part_5}/{grades[5]}")
print(f"Part 6: {part_6}/{grades[6]}")
print(f"Part 7: {part_7}/{grades[7]}")

total = sum(parts)
print(f"Your final grade is {total}")

&#169; Copyright 2022,  Department of Computational Mathematics, Science and Engineering at Michigan State University